processor/src/trace/virtual_bus/circuit/mod.rs

```diff
@@ -185,30 +210,30 @@ where
     }
 
     fn max_degree(&self) -> u32 {
-        self.degree
+        // Computed as:
+        // 1 + max(left_numerator_degree + right_denom_degree, right_numerator_degree + left_denom_degree)
+        5
     }
 
     fn evaluate(&self, query: &[E]) -> E {
-        let eval_right_numerator =
-            self.right_numerator_composer.iter().enumerate().fold(E::ZERO, |acc, (i, ml)| {
-                acc + ml.evaluate(query) * self.tensored_merge_randomness[i]
-            });
+        let LayerGatesInputs {
+            partial_left_numerator,
+            partial_right_numerator,
+            partial_left_denominator,
+            partial_right_denominator,
+        } = LayerGatesInputs::from_main_trace_query(query, &self.log_up_randomness);
+
         let eval_left_numerator =
-            self.left_numerator_composer.iter().enumerate().fold(E::ZERO, |acc, (i, ml)| {
-                acc + ml.evaluate(query) * self.tensored_merge_randomness[i]
-            });
-        let eval_right_denominator = self
-            .right_denominator_composer
-            .iter()
-            .enumerate()
-            .fold(E::ZERO, |acc, (i, ml)| {
-                acc + ml.evaluate(query) * self.tensored_merge_randomness[i]
-            });
+            inner_product(&partial_left_numerator, &self.tensored_merge_randomness);
+        let eval_right_numerator =
+            inner_product(&partial_right_numerator, &self.tensored_merge_randomness);
         let eval_left_denominator =
-            self.left_denominator_composer.iter().enumerate().fold(E::ZERO, |acc, (i, ml)| {
-                acc + ml.evaluate(query) * self.tensored_merge_randomness[i]
-            });
-        let eq_eval = self.eq_composer.evaluate(query);
+            inner_product(&partial_left_denominator, &self.tensored_merge_randomness);
+        let eval_right_denominator =
+            inner_product(&partial_right_denominator, &self.tensored_merge_randomness);
+
+        let eq_eval = query[TRACE_WIDTH];
```

Comment:
note: as you mentioned, this is fragile but should be fixed when we address the question of "automating" multiplication with evaluations of the Eq function